client/src/main/java/org/eclipse/hono/client/impl/CommandConnectionImpl.java

```diff
@@ -105,14 +116,67 @@ protected void clearState() {
                         tenantId,
                         () -> newCommandConsumer(tenantId, deviceId, commandConsumer, remoteCloseHandler))
                 .map(consumer -> {
-                    commandReceivers.put(key, consumer);
+                    commandConsumers.put(key, consumer);
                     return consumer;
                 })
                 .setHandler(result);
             }
         });
     }
 
+    /**
+     * {@inheritDoc}
+     * <p>
+     * The interval used for creating the periodic liveness check will be the maximum
+     * of the given interval length and {@link #MIN_LIVENESS_CHECK_INTERVAL_MILLIS}.
+     * 
+     */
+    public final Future<MessageConsumer> createCommandConsumer(
+            final String tenantId,
+            final String deviceId,
+            final Handler<CommandContext> commandConsumer,
+            final Handler<Void> remoteCloseHandler,
+            final long checkInterval) {
+
+        Objects.requireNonNull(tenantId);
+        Objects.requireNonNull(deviceId);
+        Objects.requireNonNull(commandConsumer);
+        if (checkInterval < 0) {
+            throw new IllegalArgumentException("liveness check interval must be > 0");
+        }
+
+        return createCommandConsumer(tenantId, deviceId, commandConsumer, remoteCloseHandler).map(c -> {
+
+            final String address = Device.asAddress(tenantId, deviceId);
+            final long effectiveCheckInterval = Math.max(MIN_LIVENESS_CHECK_INTERVAL_MILLIS, checkInterval);
+            final long livenessCheckId = vertx.setPeriodic(effectiveCheckInterval, timerId -> {
+                if (isShutdown()) {
+                    vertx.cancelTimer(timerId);
+                } else if (isConnectedInternal() && !commandConsumers.containsKey(address)) {
```

Comment:
IMHO this is a critical part of the implementation:
while the check for `commandConsumers.containsKey(address)` should work for the standard behaviour, I fear there is a danger that a recreation of a command consumer could take very long, or even hang (especially with previous connection losses).
In this case, this periodic timer could fire again and again and trigger more and more creations of command consumers, that possibly later all would succeed. Resulting in suddenly multiple command consumers, recreated out of one.

How about using something like an AtomicBoolean explicitly in this code, and check for this? By such an implementation there would be no dependency to the duration of the command consumer creation that could result in unpredictable behaviour.

Reply:
That's a very good catch :+1: 
We are already doing something similar when we create senders ...